src/context.rs

```diff
@@ -485,6 +485,7 @@ impl Context {
 
     /// Stops the IO scheduler.
     pub async fn stop_io(&self) {
+        *self.iroh.write().await = None;
```

Comment:
it would be good to call [`close`](https://docs.rs/iroh-net/latest/iroh_net/endpoint/struct.Endpoint.html#method.close) here

Reply:
summoning @flub to discuss the details of shutdown..

Reply:
Dropping the iroh endpoint will close things right away (or should, I'm not saying there are no bugs :wink:).  It will still send one UDP packet with a CONNECTION_CLOSE frame I think, but not await anything else.  The `Endpoint::close` method is async because it forces you to wait for the network, which is what it sounds like you do not want to do.

The reason to wait for the network is that this is a UDP socket rather than a TCP socket, so the kernel will not handle sending RST for us after we dropped the socket (but hey, no wait_idle and socket_reuse issues!).  So instead when you close a QUIC endpoint you send a CONNECTION_CLOSE frame for any connections which may still be open.  If you did send a CONNECTION_CLOSE frame it can still get lost, because it is UDP.  This means the remote endpoint doesn't know the connection is closed until it eventually gives up under a timeout (if configured).  So QUIC makes the CONNECTION_CLOSE ack-eliciting and when sending it you wait for the ack or resend it.  How long to wait for the resends depends on the normal connection stuff, so is based on the current RTT estimates etc.

I'm explaining all of this so you understand the consequences of not awaiting the `close`.  You're basically not being a nice citizen and leaving remote peers with dangling connections.  You also increase the likelihood that, if you re-open that UDP socket it will still receive (now garbage) traffic from all the peers that still think they can send to it.

Specifically for gossip I imagine that for the duration of the timeout you might have nodes in the (plum)tree that still think you're in this tree and basically your gossip tree is broken until the timeouts kick in and it manages to heal itself.  (I'm not very familiar with gossip, so this paragraph is less authoritative.)

(https://github.com/n0-computer/iroh/pull/1753 was misguided, we did not understand how things work back then)

Reply:
Calling `quinn::Endpoint::close` is the same as iterating over all connections and calling `Connection::close`.  If all you want to do is send one single CONNECTION_CLOSE frame on each connection then dropping the endpoint also achieves that.

`quinn::Endpoint::wait_idle` does what I described above: it waits for an acknowledgment on every CONNECTION_CLOSE frame and re-transmits any un-acknowledged frames a few times based on what the RTT estimate is for each connection.

So calling `iroh_net::Endpoint::close().await` **will** wait for all of this network stuff to happen and is not going to be fast but wait on network traffic.  (You may not notice that in your testing because you need to have open connections and packet loss for `iroh_net::Endpoint::close` to be slow)